### ISSUE TITLE
General linear model.ipynb

### ISSUE THREAD

**Issue by JNPauli** [JNPauli/Mental-image-decoding#8]
Hello! 

For now, I did quite a bit of work on the GLM. However, I am not sure if the steps made until now are correct. Could one of you please check, if the analysis process is correct so far? 

Especially because "plot_stat_map" and "plot_glass_brain" look.. weird. Like something was not correct.

**Comment by PeerHerholz** [comment]
Hi,

for subsequent issues, please make sure to link to the respective files, as e.g. the GLM notebook is not on the `main` but `update` `branch`.

As far as I can tell, it looks ok. I would maybe suggest the following:

- [x] go through [this part of the nilearn docs](https://nilearn.github.io/stable/auto_examples/04_glm_first_level/plot_first_level_details.html#sphx-glr-auto-examples-04-glm-first-level-plot-first-level-details-py) and evaluate your `GLM` respectively
- [x] write down what you would expect from the outcomes (how "should" things look like)
- [x] include more than one run in the `GLM`, ie generate a `design matrix` for `run-02`, etc. and add them to the `GLM` (this is e.g. outlined [here](https://nilearn.github.io/stable/auto_examples/04_glm_first_level/plot_fiac_analysis.html#sphx-glr-auto-examples-04-glm-first-level-plot-fiac-analysis-py)) to get "more power" for a given `contrast`, ie get the responses to a `category` accumulated/pooled across `runs` 

HTH, cheers, Peer

**Comment by JNPauli** [comment]
Thank you so much!!

So I worked through the steps and pushed the updated notebook to the update branch.

However, I am not sure if I understood this correctly. Am I supposed to compute 5 different matrices and include them all in the SAME glm.fit() function? Or run 5 seperate glm.fit() This was not really clear to me, since [some resources](https://nilearn.github.io/stable/auto_examples/02_decoding/plot_haxby_glm_decoding.html#run-the-glm-on-data-from-each-session) solved it different. 

I would appreciate a brief evaluation of my current notebook a lot, because I am not sure, wether I did this correctly or not.
I provided both approaches in the current version of the notebook. I hope my issue is clear.

Janos

**Comment by PeerHerholz** [comment]
Hi there,

thanks for the updates. (NB: the `checkboxes` can be `ticked` to indicate if the respective point was addressed or even better: they can be linked to `commits` and `PR`s.)

> However, I am not sure if I understood this correctly. Am I supposed to compute 5 different matrices and include them all in the SAME glm.fit() function? Or run 5 seperate glm.fit() This was not really clear to me, since [some resources](https://nilearn.github.io/stable/auto_examples/02_decoding/plot_haxby_glm_decoding.html#run-the-glm-on-data-from-each-session) solved it different.

As note above, all `runs` and `design matrices` should be submitted to one `GLM`, ie the first option you outlined. The other option would run one separate `GLM` for each `run` which would make it harder to accumulate and pool across `runs`. Re the `notebook`: could you please implement this respectively and add a `plot`, as well as `evaluate` the resulting `design matrix`  as done before? NB: maybe try adding multiple `design matrices` to one `GLM` as outlined [here](https://nilearn.github.io/stable/auto_examples/04_glm_first_level/plot_fiac_analysis.html#prepare-data-and-analysis-parameters). 

HTH, cheers, Peer

**Comment by JNPauli** [comment]
Hi!

Thanks a lot! This helps a lot. So should I add 5 different plots for all 5 desgin matrices?

> NB: maybe try adding multiple design matrices to one GLM as outlined [here](https://nilearn.github.io/stable/auto_examples/04_glm_first_level/plot_fiac_analysis.html#prepare-data-and-analysis-parameters).

Regarding this: I actually did do this in my notebook, or did I execute it incorrectly? This is the part of the respective notebook:
I just realized though, that I forgot to delete the "z_maps_all" bit of the loop. But I am supposed to do it just like outlined below, right?

`os.chdir('/mnt/c/Users/janos/git/sub01-ses-01-imagery')

runs = [0,1,2,3,4]
matrix = []
for pos, x in enumerate(runs):
    glm.fit(fmri_img_all_runs[pos],regressors_of_interest[pos],reg_no_interest_all[pos])
    matrix.append(glm.design_matrices_[0])`

and

glm = glm.fit(fmri_img_all_runs, design_matrices=matrix)

I have to fit two glm, because I need to create the respective design matrices first.. This is why I approached it this way. In the nilearn example, the design matrices are already precomputed and I did not know how to compute them other than in the glm.

**Comment by PeerHerholz** [comment]
Hi,

> Thanks a lot! This helps a lot. So should I add 5 different plots for all 5 desgin matrices?

It should be possible to create one plot that includes all `runs`. 

> I just realized though, that I forgot to delete the "z_maps_all" bit of the loop. But I am supposed to do it just like outlined below, right?

In this case, you're obtaining the `design matrices` after  a `GLM` was already `fit`ted to the `data`.  I would not recommend this approach. Maybe try following this [section/example](https://nilearn.github.io/stable/auto_examples/04_glm_first_level/plot_design_matrix.html#examples-of-design-matrices), ie creating `design matrices` first and then providing a list of them as input to the `GLM`.

HTH, cheers, Peer

**Comment by JNPauli** [comment]
Hi,

alright! I calculated the design matrices just like in the example.and pushed it to the update branch.
So I guess this is it? I saved the z-maps for each contrast. 
According to the nilearn [glm to decoding](https://nilearn.github.io/stable/auto_examples/02_decoding/plot_haxby_glm_decoding.html#build-the-decoding-pipeline) example, the z_maps are the input to the SVM. Can I now proceed with the SVM?

Regarding the contrast plotting I was unsure, wether I should plot every single contrast or just a few to illustrate.

> It should be possible to create one plot that includes all runs.

But the plot would be kind of messy, right? I mean I could create 5 subplots, but all categories over time for all runs in one figure seems rather uninformative to me. Or am I missing something obvious here?

Thanks a lot for your help

Janos

**Comment by PeerHerholz** [comment]
Hi,

> alright! I calculated the design matrices just like in the example.and pushed it to the update branch.

just a quick note re terminology: `design matrices` are not calculated but defined/assembled. 

> So I guess this is it? I saved the z-maps for each contrast.
According to the nilearn [glm to decoding](https://nilearn.github.io/stable/auto_examples/02_decoding/plot_haxby_glm_decoding.html#build-the-decoding-pipeline) example, the z_maps are the input to the SVM. Can I now proceed with the SVM

You could, but I would strongly advise evaluating your results beforehand. Specifically, so far the `contrasts`/`z-maps` were not checked via respective `plots` and/or the `model performance` assessed, e.g. via `r2` or so. This basically also relates to the second point, ie the one that was not checked yet: are the results as expected, that was what kind of results would one expect for `visual imagery` based on prior literature? There are also automated `meta-analysis` like [neurosynth](https://www.neurosynth.org/) and [neuroquery](https://neuroquery.org/query?text=putamen) that can help with.

> Regarding the contrast plotting I was unsure, wether I should plot every single contrast or just a few to illustrate.

>> It should be possible to create one plot that includes all runs.

> But the plot would be kind of messy, right? I mean I could create 5 subplots, but all categories over time for all runs in one figure seems rather uninformative to me. Or am I missing something obvious here?

This was not referring to the `contrasts` but the `design matrices`. As with the `design matrix` for one `run`, the appended/accumulated `design matrices` should also be `plotted` and checked. [Here's an example](https://neurostars.org/t/first-level-modeling-in-nilearn-conditions-come-from-different-runs/19816/5?u=peerherholz) of how this would like. 

However, as mentioned above also all `contrasts` (of interest) should be `plotted`, ideally in two ways: each `category` for per `run` (e.g. `cat_run_1`, `dog_run_1`, etc.) and each `category` across `runs` (e.g. `cat_all_runs`, `dog_all_rungs`).

HTH, cheers, Peer

**Comment by JNPauli** [comment]
Hi,

thank you!

So just to be clear: I have to

- [x] Check the model performace via R^2, so probably just follow the steps of the [course notebook](https://peerherholz.github.io/Cog_Com_Neuro_ML_DL/introduction/notebooks/neuroscience/statistical_analyses_MRI.html#evaluating-models)
- [x] Indicate what results I would actually expect
- [x] Plot the design matrices
- [x] Plot all contrasts of interest 

Regarding the contrasts:
I need to plot all 26 contrasts (=26 regressors of interest) and do that per run and across runs.
So in order to do that, I need to calculate a z-map for each run and plot those. (category per run)
Could I just plot the z-maps I calculated with my in the lastet commit for the "across runs" contrast? I mean that is basically what I computed there, right?

Thanks a lot

Janos

**Comment by PeerHerholz** [comment]
Hi,

sounds about right.

> Regarding the contrasts:
I need to plot all 26 contrasts (=26 regressors of interest) and do that per run and across runs.
So in order to do that, I need to calculate a z-map for each run and plot those. (category per run)
Could I just plot the z-maps I calculated with my in the lastet commit for the "across runs" contrast? I mean that is basically what I computed there, right?

Are there 26 `categories` in each `run`? If so, I would the following:

- `plot` each `category`  per `run`
- `plot` each `category` across `runs`

which means you should have 6 `plots` per `category` (as you have 5 `runs` and one across all `runs`). The `across runs` `contrast` would give you the latter but not the former, as it is accumulated/pooled across `runs`.

HTH, cheers, Peer

**Comment by JNPauli** [comment]
Hi,

this would be equal to 6*26 plots. This is quite a lot. Should I create a new notebook for this? I guess this would blow up my notebook. Or is it maybe sufficient to plot one contrast per run and one contrast across runs?

Or maybe just plot after the other: So contrast dog-run(i)...contrast dog-run(j) and dog-run (across runs) and then remove the existing plot and repeat the steps with contrast-cat (...)?

What should I aim for? Look if the activity is similiar in respect to results from neurosynth?

Thanks a lot

Janos

**Comment by PeerHerholz** [comment]
Hi,

> this would be equal to 6*26 plots. This is quite a lot. Should I create a new notebook for this? I guess this would blow up my notebook. Or is it maybe sufficient to plot one contrast per run and one contrast across runs?

> Or maybe just plot after the other: So contrast dog-run(i)...contrast dog-run(j) and dog-run (across runs) and then remove the existing plot and repeat the steps with contrast-cat (...)?

That's up to you and how you want to implement it. I would argue that one should check all the data carefully to prevent problems further down the analysis stream. Quality control of every step is quite important and commonly forgotten. However, that's just my two cents.

> What should I aim for? Look if the activity is similiar in respect to results from neurosynth?

Not only but definitely a good start. The basic idea would be to evaluate if the maps you see "make sense" given the `paradigm` of the study and the `model` you applied, as well as/specifically integrating results from prior research. For example, based on earlier studies: what would expect the map to look like, where and how is "activity" distributed? Is this stable across `runs`, `participants` and `categories`? 

HTH, cheers, Peer

**Comment by JNPauli** [comment]
Hi,

thanks!

I created the R square plots and also created the across runs contrasts. I added an example from neurosynth with similar coordinates from a given contrast plot.

Am I doing it the right way? There are similarities in the activation pattern, but they do not 100% overlap. But f.e. in the seconds contrast plot, in the saggital cut activation in the visual cortex can be seen. This is not according to neurosynth images, but makes sense in theory.

Thanks a lot

Janos 

EDIT:

I now completed all the of the bullet points. We already discussed some parts on discord. Could I still get a review for the notebook chapter "4.0" - "4.4"?
